Fix Usage of IOObjects on Mac

LumioseSil opened the pull request:
#### 570db2bde74d80b5aba9333e2fff3c835dedd80f
<pre>
Fix Usage of IOObjects on Mac
<a href="https://bugs.webkit.org/show_bug.cgi?id=251859">https://bugs.webkit.org/show_bug.cgi?id=251859</a>

Reviewed by Brent Fulgham.

IOMasterPort is deprecated, but still used up until Big Sur, so let&apos;s
add a check for IOMainPort and then use it if the OS has it.

Additionally, there is an unnecessary retain + release before
IOServiceOpen, so let&apos;s remove that.

*Source/WTF/wtf/PlatformHave.h: Add check for IOMainPort
*Source/WebCore/platform/graphics/mac/GraphicsChecksMac.cpp: Rename
masterPort to mainPort and remove pointless initialization. Also remove
unneeded IOObjectRetain and IOObjectRelease calls in the same exact
codepath. Replace calls from IOMasterPort to IOMainPort
*Tools/DumpRenderTree/mac/LayoutTestHelper.m: Rename masterPort to
mainPort. Replace calls from IOMasterPort to IOMainPort

Canonical link: <a href="https://commits.webkit.org/260226@main">https://commits.webkit.org/260226@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42d330b5aa941791a39a61caef4579ec970a7a0c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107539 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16599 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40442 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116703 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116122 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111435 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18026 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7843 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99707 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113298 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13613 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96797 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41270 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95522 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28423 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96874 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9604 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29776 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/96289 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7604 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10262 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6677 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/30818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15761 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49357 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/105144 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7068 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11823 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26072 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -971,6 +971,7 @@
 #define HAVE_CFNETWORK_NSURLSESSION_HSTS_WITH_UNTRUSTED_ROOT 1
 #define HAVE_NSURLSESSION_TASK_DELEGATE 1
 #define HAVE_AUDIO_OBJECT_PROPERTY_ELEMENT_MAIN 1
+#define HAVE_MAIN_PORT 1
```
Based on the API declaration for `IOMainPort`, I think you should do something like this:

```
#if (PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 120000) \
    || PLATFORM(IOS_FAMILY)
#define HAVE_IOKIT_MAIN_PORT 1
#endif
```

AZero13 replied:
This matches exactly the ifdef above it though

brentfulgham replied:
We build WebKit on the current OS (Ventura, 130000) and two down-level OSes (Monterey, 120000 and Big Sur, 110000). Big Sur does not have IOKitMain, and so your change will cause a build failure on that platform. Please use the definition I suggested to you.

brentfulgham commented on Source/WebCore/platform/graphics/mac/GraphicsChecksMac.cpp:
```diff
@@ -43,20 +43,25 @@ enum {
-    mach_port_t masterPort = MACH_PORT_NULL;
+    mach_port_t mainPort;
 
+#if HAVE(MAIN_PORT)
```
`#if HAVE(IOKIT_MAIN_PORT)`

brentfulgham commented on Source/WebCore/platform/graphics/mac/GraphicsChecksMac.cpp:
```diff
@@ -65,9 +70,7 @@ static io_connect_t attachToAppleGraphicsControl()
         return IO_OBJECT_NULL;
 
     io_connect_t dataPort;
-    IOObjectRetain(serviceObject);
```
I'm not confident it is safe to remove the Retain/Release pair. Can you explain why this is safe?

AZero13 replied:
Reference count goes up by one and down by one. 

The tools folder has a code snippet like this without the retain and release. 

If the retain was needed, the release should not be there.

brentfulgham replied:
I did some looking, and do not see other places where a Retain/Release is needed around IOServiceOpen, so this change is probably good.

brentfulgham commented on Tools/DumpRenderTree/mac/LayoutTestHelper.m:
```diff
@@ -223,15 +223,23 @@ static void simpleSignalHandler(int sig)
-    if (kernResult != KERN_SUCCESS)
+    mach_port_t mainPort;
+
+#if HAVE(MAIN_PORT)
```
Ditto `HAVE(IOKIT_MAIN_PORT)`

brentfulgham commented on Tools/DumpRenderTree/mac/LayoutTestHelper.m:
```diff
@@ -242,15 +250,16 @@ static void lockDownDiscreteGraphics(void)
 
     // This call stalls until the graphics device lock is granted.
+    kern_return_t kernResult;
     kernResult = IOServiceOpen(serviceObject, mach_task_self(), 1, &permanentLockDownService);
```
I would just add the `kern_return_t` to line 254.

brentfulgham commented on Tools/DumpRenderTree/mac/LayoutTestHelper.m:
```diff
@@ -242,15 +250,16 @@ static void lockDownDiscreteGraphics(void)
     }
 
-    kernResult = IOObjectRelease(serviceObject);
+    kernResult = IOServiceClose(serviceObject);
```
Ordinarily this would be the right thing to do. But you missed the comment on line 240:

```
// We're intentionally leaking this io_connect in order for the process to stay locked to discrete graphics
// for the lifetime of the service connection.
```

By calling `IOObjectRelease` we release the bulk of the object memory, while holding the connection open for the life of the process. If you call `IOServiceClose`, it will shut down that connection removing the hold on the discrete GPU.

brentfulgham reviewed (changes requested):
Much of this change is good, but I suggest that we use the version-check for HAVE_MAIN_PORT, and the switch to `IOServiceClose` would likely break the ability of the process to choose discrete GPU.

brentfulgham commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -971,6 +971,7 @@
 #define HAVE_CFNETWORK_NSURLSESSION_HSTS_WITH_UNTRUSTED_ROOT 1
 #define HAVE_NSURLSESSION_TASK_DELEGATE 1
 #define HAVE_AUDIO_OBJECT_PROPERTY_ELEMENT_MAIN 1
+#define HAVE_IOKIT_MAIN_PORT 1
```
We build WebKit on the current OS (Ventura, 130000) and two down-level OSes (Monterey, 120000 and Big Sur, 110000). Big Sur does not have IOKitMain, and so your change will cause a build failure on that platform. Please use the definition I suggested to you in an earlier review comment.

AZero13 replied:
#if (PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 120000) \
    || PLATFORM(IOS_FAMILY) is literally the condition above all this though.

And the EWS is on a Big Sur build, which compiles as seen by the results.

AZero13 replied:
Please please please look at the code surrounding my definition.

Please. 

It says right there it's only if it's >= 12000

brentfulgham reviewed (changes requested):
The current code will break builds on macOS Big Sur, so we can't accept the patch as-is.

brentfulgham reviewed (approved):
Change looks good now. Thanks!

LumioseSil committed:
Fix Usage of IOObjects on Mac
https://bugs.webkit.org/show_bug.cgi?id=251859

Reviewed by Brent Fulgham.

IOMasterPort is deprecated, but still used up until Big Sur, so let's
add a check for IOMainPort and then use it if the OS has it.

Additionally, there is an unnecessary retain + release before
IOServiceOpen, so let's remove that.

*Source/WTF/wtf/PlatformHave.h: Add check for IOMainPort
*Source/WebCore/platform/graphics/mac/GraphicsChecksMac.cpp: Rename
masterPort to mainPort and remove pointless initialization. Also remove
unneeded IOObjectRetain and IOObjectRelease calls in the same exact
codepath. Replace calls from IOMasterPort to IOMainPort
*Tools/DumpRenderTree/mac/LayoutTestHelper.m: Rename masterPort to
mainPort. Replace calls from IOMasterPort to IOMainPort

Canonical link: https://commits.webkit.org/260226@main
